src/Pos/Explorer/BListener.hs

```diff
@@ -0,0 +1,255 @@
+-- Callbacks on application and rollback.
+
+module Pos.Explorer.BListener
+       ( runExplorerBListener 
+       -- * Instances
+       -- ** MonadBListener (ExplorerBListener m)
+       ) where
+
+import           Universum
+
+import           Control.Monad.Trans.Identity (IdentityT (..))
+import           Data.Coerce                  (coerce)
+import qualified Data.List.NonEmpty           as NE
+import qualified Data.Map                     as M
+import qualified Ether
+import           System.Wlog                  (WithLogger)
+
+import           Mockable                     (MonadMockable)
+import           Pos.Block.BListener          (MonadBListener (..))
+import           Pos.Block.Core               (Block)
+import           Pos.Block.Types              (Blund)
+import           Pos.Core                     (HeaderHash, difficultyL,
+                                               epochIndexL, getChainDifficulty,
+                                               headerHash)
+import           Pos.DB.BatchOp               (SomeBatchOp (..))
+import           Pos.DB.Class                 (MonadDBRead, MonadRealDB)
+import           Pos.Explorer.DB              (Page, Epoch)
+import qualified Pos.Explorer.DB              as DB
+import           Pos.Ssc.Class.Helpers        (SscHelpersClass)
+import           Pos.Util.Chrono              (NE, NewestFirst (..),
+                                               OldestFirst (..))
+
+----------------------------------------------------------------------------
+-- Declarations
+----------------------------------------------------------------------------
+
+data ExplorerBListenerTag
+
+type ExplorerBListener = Ether.TaggedTrans ExplorerBListenerTag IdentityT
+
+-- Kind of... Empty and devoid of meaning...
+runExplorerBListener :: ExplorerBListener m a -> m a
+runExplorerBListener = coerce
+
+-- Explorer implementation for usual node. Combines the operations.
+instance ( MonadRealDB m
+         , MonadDBRead m
+         , MonadMockable m
+         , WithLogger m
+         )
+         => MonadBListener (ExplorerBListener m) where
+    onApplyBlocks     blunds = do
+        epochBlocks <- onApplyEpochBlocksExplorer blunds
+        pageBlocks  <- onApplyPageBlocksExplorer blunds
+        pure $ SomeBatchOp [epochBlocks, pageBlocks]
+
+    onRollbackBlocks  blunds = do
+        epochBlocks <- onRollbackEpochBlocksExplorer blunds
+        pageBlocks  <- onRollbackPageBlocksExplorer blunds
+        pure $ SomeBatchOp [epochBlocks, pageBlocks]
+
+-- onGeneralEpochBlocksExplorer is viable, since the only thing that needs 
+-- to be defined are reversed operations (adding and removing)
+
+----------------------------------------------------------------------------
+-- Apply
+----------------------------------------------------------------------------
+
+-- For @EpochBlocks@
+onApplyEpochBlocksExplorer
+    :: forall ssc m .
+    ( SscHelpersClass ssc
+    , WithLogger m
+    , MonadRealDB m
```

Comment:
Actually I think `MonadRealDB` is not needed in this whole module, but I am not sure.